---
title: "scFM: Bayesian segmented Gaussian copula factor model for single-cell sequencing data"
author: "Junsouk Choi"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{scFM}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r setup, include = FALSE}
knitr::opts_chunk$set(
  echo = TRUE,
  message = FALSE,
  warning = FALSE,
  collapse = TRUE,
  comment = "#>",
  eval = TRUE,
  cache = FALSE,
  tidy = TRUE,
  tidy.opts = list(width.cutoff = 60),
  fig.width = 9,
  fig.height = 6,
  fig.align = "center"
)
```


# Introduction

The 'scFM' package implements the Segmented Gaussian Copula Factor Model (scFM), a Bayesian latent factor model designed for single-cell RNA sequencing (scRNA-seq) data, as introduced by Choi et al. (2024). As described in paper, "The proposed segmented Gaussian copula factor model (scFM) handles both zero-inflated and skewed counts in scRNA-seq data by introducing a segmentation of the empirical marginal distribution.” (Section 1, p.1)

# Data Preparation

This vignette uses the scFM package, which provides the core functionality for the Segmented Gaussian Copula Factor Model and includes the built-in dataset used in this analysis. The paper describes the dataset: "We use the GM12878 cell line (LCL) dataset, which consists of 5,135 cells and 100 genes after preprocessing" (Section 5.1, p.18).

```{r load-data}
# Load the scFM package, which provides the core functionality and built-in dataset
library(scFM)
set.seed(7)
# Load the built-in LCL dataset
data(LCL)
x = LCL
# Check dimensions of the dataset
dim(x)
```

# Model Initialization and Prior Setup

```{r load-packages}
# load necessary packages
library(tmvnsim)
library(tmvtnorm)
library(GIGrvg)
library(actuar)
library(ggplot2)
library(SingleR)
library(celldex)
library(tibble)
```

## Hyperparameter Settings

```{r set-priors}
# set hyperparameters
priors = list()
priors$Sigma = c(0.1, 0.1)
priors$Phi   = 0.5
```


## Empirical Marginal Transformation

We transform the marginal distributions of observed counts to approximate standard normals using empirical CDFs

```{r init-Z-delta}
# initialize parameters Z, delta
starting = list()
n = nrow(x)
p = ncol(x)
k = 8
ecdf.scale = n / (n + 1)
eFx   = apply(x, 2, ecdf)
eFx   = lapply(eFx, function(x){  function(y) ecdf.scale * x(y)  })
eFxx  = Map(function(f, x) do.call(f, list(x)), eFx, plyr::alply(x, 2))
starting$Z     = matrix(unlist(lapply(eFxx, function(pr) qnorm(pr))), n, p)
starting$delta = matrix(0, p, 2)
starting$delta[ , 1] = qnorm(colMeans(x == 0))
starting$delta[ , 2] = qnorm(colMeans(x <= 1))
```


## Initial Factor Estimates via Factor Analysis

We initialize the loadings and scores using standard factor analysis on the latent Z.

```{r init-Lambda-U}
# initialize Lambda and U from factor analysis
out0 = factanal(starting$Z, factors = k, scores = "regression")
starting$Lambda = as.matrix(out0$loadings)
starting$U = out0$scores
```


## Sampling Prior Hyperparameters

we place a column-wise Dirichlet–Laplace (DL) prior on the loading matrix $\Lambda$, where each element $\lambda_{jh}$ $\sim$ $\text{DE}(\tau \phi_h)$, and the global and local scales $\tau$ and $\phi$ = $(\phi_1, \dots, \phi_k)$ are sampled from their full conditionals

```{r sampling-hyperparamters}
# Sampling Hyperparameters for Initialization
Sigma = diag(1, p)
Phi = rep(1 / k, k)
tau = 1
Psi = matrix(rexp(p * k, rate = 0.5), p, k)

for (iter in 1 : 1000) {
   Sigma[diag(rep(TRUE, p))] = 1 / rgamma(p, shape = priors$Sigma[1] + rep(n / 2, p), 
                                          rate = priors$Sigma[2] + 0.5 * colSums((starting$Z - tcrossprod(starting$U, starting$Lambda))^2))
   
   H = rep(NA, k)
   for (l in 1 : k) {
      H[l] = rgig(1, lambda = priors$Phi - p, chi = 2 * sum(abs(starting$Lambda[ , l])), psi = 1)
   }
   Phi = H / sum(H)
   tau = rgig(1, lambda = k * (priors$Phi - p), chi = 2 * sum(t(abs(starting$Lambda)) / Phi), psi = 1)
   Psi_tilde = rinvgauss(p * k, mean = c(t(Phi * t(tau / abs(starting$Lambda)))))
   Psi = matrix(1 / Psi_tilde, p, k)
}

starting$Sigma = Sigma
starting$Phi   = Phi
starting$tau   = tau
starting$Psi   = Psi
```

# Posterior Inference

## Gibbs Sampling

We implement a blocked Gibbs sampler for the scFM model to obtain posterior draws from the param    eters

```{r scFM}
nmcmc   = 100
nburnin = 50
p.time = proc.time()
out    = gibbs_scFM(x, k, m = 1, starting, priors, nmcmc = nmcmc, nburnin = nburnin)
proc.time() - p.time
```

We compute posterior means of the factor loadings and scores by averaging over MCMC samples.

```{r posterior-means}
# Compute posterior means
loading_est = apply(out$Lambda, c(1, 2), mean) 
score_est = apply(out$U, c(1, 2), mean) 
```


## Latent Factor Selection

To identify significant latent factors, we perform k-means clustering on the columns of Lambda.

```{r identify-significant-factors}
# Identify significant factors using k-means clustering
sig_facto = matrix(0, dim(out$Lambda)[2], dim(out$Lambda)[3])
for (i in 1 : dim(out$Lambda)[3])
{
   kmeans_lambda = kmeans(t(out$Lambda[ , , i]), centers = 2, nstart = 10)
   sig_facto[which(kmeans_lambda$cluster == which.max(rowSums(kmeans_lambda$centers^2))), i] = 1
}
table(colSums(sig_facto))
rowMeans(sig_facto)   # only the first two factors are significant

score_est_all = score_est
score_est = score_est[ , 1 : 2]
```

# Downstream analysis

We use the estimated factor scores to cluster cells and visualize latent structure.

## Clustering and Visualization

We try to explore the latent structure using the estimated factor scores, following the example in the paper. "We use them to identify distinct cellular subtypes or states in LCL. The scores of the two selected factors serve as a low-dimensional representation of the high-dimensional gene expression measurements in scRNA-seq data.” (Section 5.3, p.21)

To decide how many clusters to use, we draw a scree plot based on within-cluster sum of squares.

"For simplicity, we use the k-means algorithm to cluster the latent factor scores, where the number of clusters is determined by the elbow method." (Section 5.3, p.21)

We then visualize the clustering result using the first two factors.


```{r scree-plot}
# Scree plot to determine number of clusters
n_clusters = 10
wss = numeric(n_clusters)
for (i in 1 : n_clusters)
{
   # fit the model: km.out
   km.out = kmeans(score_est, centers = i, nstart = 100, iter.max = 10000)
   # save the within cluster sum of squares
   wss[i] = km.out$tot.withinss
}

# produce a scree plot to determine the number of the underlying cell types
wss_df = tibble(clusters = 1 : n_clusters, wss = wss)
scree_plot = ggplot(wss_df, aes(x = clusters, y = wss, group = 1)) + 
   geom_point(size = 4) + 
   geom_line() +
   scale_x_continuous(breaks = c(2, 4, 6, 8, 10)) +
   xlab('Number of clusters')
scree_plot
```

We visualize the clustering results on the first two significant factors identified by our shrinkage prior.

```{r clustering}
# Apply k-means clustering with chosen k = 3
out_cell_types = kmeans(score_est, centers = 3, nstart = 100, iter.max = 10000)
df = data.frame(facto1 = score_est[ , 1], facto2 = score_est[ , 2], Cluster = factor(out_cell_types$cluster))
plot_cell_types = ggplot(df[-4148, ], aes(facto1, facto2, color = Cluster)) + geom_point(alpha = 0.3) + labs(x = "Factor 1", y = "Factor 2")
plot_cell_types
```

## Top 10 Genes by Factors

We look at the absolute values of the loadings to find out which genes are most related to each factor.

"Figure 3 illustrates the absolute loadings of all genes on these two significant factors, providing insight into a potential biological interpretation for each estimated factor." (Section 5.2, p.19)

For factor 1, the top genes include MKI67, CENPF, etc., which are related to the cell cycle.
"Therefore, we call factor 1 the cell cycle factor." (Section 5.2, p.19)

For factor 2, the top genes include MIR155HG, FSCN1, etc., which are related to immune function.
"Therefore, we call factor 2 the immune cell factor." (Section 5.2, p.19)

To check this, we also look at the average expression of these genes in each cluster.


```{r top-10, fig.width = 12, fig.height = 6}
# Plot top 10 genes by loadings on each factor
rownames(loading_est) = colnames(x)

order1 = order(abs(loading_est[ , 1]), decreasing = TRUE)
df = data.frame(gene = factor(colnames(x), levels = colnames(x)[order1]), value = abs(loading_est[ , 1]))
loading1 = ggplot(df, aes(x = gene, y = value)) + geom_bar(stat = "identity") + scale_x_discrete(guide = guide_axis(n.dodge = 5)) + labs(x = "Genes", y = "Absolute loadings on factor 1")
loading1

order2 = order(abs(loading_est[ , 2]), decreasing = TRUE)
df = data.frame(gene = factor(colnames(x), levels = colnames(x)[order2]), value = abs(loading_est[ , 2]))
loading2 = ggplot(df, aes(x = gene, y = value)) + geom_bar(stat = "identity") + scale_x_discrete(guide = guide_axis(n.dodge = 5)) + labs(x = "Genes", y = "Absolute loadings on factor 2")
loading2

# top 10 significant factor loadings
sort(loading_est[rank(abs(loading_est[ , 1])) > 90, 1], decreasing = TRUE)
sort(loading_est[rank(abs(loading_est[ , 2])) > 90, 2], decreasing = TRUE)

round(colMeans(x[ , rank(abs(loading_est[ , 1])) > 90]), digits = 3)
round(colMeans(x[ , rank(abs(loading_est[ , 2])) > 90]), digits = 3)

# average expression of genes corresponding to significant loadings across different cell subtypes
round(colMeans(x[out_cell_types$cluster == 1, rank(abs(loading_est[ , 1])) > 90]), digits = 3)
round(colMeans(x[out_cell_types$cluster == 1, rank(abs(loading_est[ , 2])) > 90]), digits = 3)
round(colMeans(x[out_cell_types$cluster == 2, rank(abs(loading_est[ , 1])) > 90]), digits = 3)
round(colMeans(x[out_cell_types$cluster == 2, rank(abs(loading_est[ , 2])) > 90]), digits = 3)
round(colMeans(x[out_cell_types$cluster == 3, rank(abs(loading_est[ , 1])) > 90]), digits = 3)
round(colMeans(x[out_cell_types$cluster == 3, rank(abs(loading_est[ , 2])) > 90]), digits = 3)

```

## Cluster Annotation

We additionally annotate the clusters using SingleR.

"We further validate the biological identity of the clusters by annotating them using SingleR, a reference-based method that compares expression profiles of clustered cells to reference cell types." (Section 5.3, p.22)

```{r annotate}
# Annotate clusters using SingleR
hpca.se = HumanPrimaryCellAtlasData()
hpca.se
annotate_clusters = SingleR(test = t(x), ref = hpca.se, clusters = out_cell_types$cluster, labels = hpca.se$label.main)
annotate_clusters
```
